### PR TITLE
conversational retrieval agent

### PR DESCRIPTION
**Pull request by mcantillon21**
focused now on the "expert" mode of the previous implantation: a custom tools conversational retrieval agent with three indexes: 

1) Python API Reference + Documentation
2) Code base
3) URL Source Embedding Search

This provides a more holistic search than retrieving from one single index.

### CODE REVIEW

File: _scripts/evaluate_chains_agent.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+def search(inp: str, index_name: str, callbacks=None) -> str:
+    client = weaviate.Client(url=WEAVIATE_URL, auth_client_secret=weaviate.AuthApiKey(api_key=WEAVIATE_API_KEY))
+
+    weaviate_client = Weaviate(
+        client=client,
+        index_name=index_name,
+        text_key="text",
+        embedding=OpenAIEmbeddings(chunk_size=200),
+        by_text=False,
+        attributes=["source"] if not index_name == "LangChain_agent_sources" else None,
```
Comment (jacoblee93) [inline _scripts/evaluate_chains_agent.py:51]: Different index name with sources?

File: _scripts/evaluate_chains_agent.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    system_message = SystemMessage(
+            content=(
+                "You are a helpful chatbot who is tasked with answering questions about LangChain."
+                "Answer the following question as best you can."
+                "Be inclined to include CORRECT Python code snippets if relevant to the question. If you don't know the answer, just say you don't know."
+                "You have access to a LangChain knowledge bank retriever tool for your answer."
+            )
+    )
+    
+    if chat_history:
```
Comment (jacoblee93) [inline _scripts/evaluate_chains_agent.py:95]: Conditional is necessary here?

File: chat-langchain/app/components/ChatWindow.tsx [hunk 3]
```diff
@@ -27,22 +27,22 @@ export function ChatWindow(props: {
...
 
   const {
     apiBaseUrl,
     placeholder,
     titleText = "An LLM",
   } = props;
 
+  
   const sendMessage = async (message?: string) => {
+    e.preventDefault();
```
Comment (jacoblee93) [inline chat-langchain/app/components/ChatWindow.tsx:45]: This doesn't look like it should be there?

File: chat-langchain/app/components/ChatWindow.tsx [hunk 4]
```diff
@@ -267,19 +254,10 @@ export function ChatWindow(props: {
           onKeyDown={(e) => {
             if (e.key === "Enter" && !e.shiftKey) {
               e.preventDefault();
-              sendMessage();
+              sendMessage(e);
```
Comment (jacoblee93) [inline chat-langchain/app/components/ChatWindow.tsx:257]: This doesn't match the method signature above

File: main.py [hunk 5]
```diff
@@ -37,129 +43,122 @@
...
+                "For another example, given the following input question:\n"
+                "-----START OF EXAMPLE INPUT QUESTION-----\n"
+                "How can I use vLLM to run my own locally hosted model? \n"
+                "-----END OF EXAMPLE INPUT QUESTION-----\n"
+                "Your research flow should be:\n"
+                "1. Query your search tool for information on 'vLLM' to get as much context as you can about it. \n"
+                "2. Answer the question as you now have enough context."
+    ))
+    
     if chat_history:
```
Comment (jacoblee93) [inline main.py:116]: Same question as above - I think you can always just pass a memory in and if there's no chat messages it'll just populate `MessagesPlaceholder` with nothing

File: _scripts/evaluate_chains_agent.py (hunk 1 above)
Comment (mcantillon21) [reply _scripts/evaluate_chains_agent.py:51]: yes now we have three indexes

File: main.py (hunk 5 above)
Comment (mcantillon21) [reply main.py]: fixed

File: _scripts/evaluate_chains_agent.py (hunk 2 above)
Comment (mcantillon21) [reply _scripts/evaluate_chains_agent.py]: yup changed

**Pull request merged**